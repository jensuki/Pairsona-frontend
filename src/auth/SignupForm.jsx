import { useState, useContext } from 'react';
import { useNavigate } from 'react-router-dom';
import UserContext from '../context/UserContext';
import ReactGoogleAutocomplete from 'react-google-autocomplete'; // for autocomplete dropdown
import Loading from '../components/Loading';
import '../styles/Forms.css';

const SignupForm = () => {
    const { signup } = useContext(UserContext);
    const navigate = useNavigate();

    const [formData, setFormData] = useState({
        username: '',
        password: '',
        firstName: '',
        lastName: '',
        email: '',
        birthDate: '',
        location: '',
        bio: '',
        profilePic: null // for file upload
    });

    const [picPreview, setPicPreview] = useState(null);
    const [error, setError] = useState(null);
    const [loading, setLoading] = useState(false);

    // handle form input change
    const handleChange = (e) => {
        const { name, value } = e.target;

        if (name === 'bio' && value.length > 100) {
            setError('Bio cannot exceed 100 characters');
        } else {
            setError(null); // clear the error
        }

        setFormData((fData) => ({
            ...fData,
            [name]: value
        }));
    }

    // handle location suggestion autocomplete
    const handleLocationChange = (city) => {
        setFormData((fData) => ({
            ...fData,
            location: city.formatted_address // store formatted address
        }));
    };

    // handle image file upload + create preview
    const handleFileUpload = (e) => {
        const file = e.target.files[0];
        if (file) {
            setFormData((fData) => ({
                ...fData,
                profilePic: file,
            }));

            // generate preview
            setPicPreview(URL.createObjectURL(file));
        }
    };

    // handle signup form submission
    const handleSubmit = async (e) => {
        e.preventDefault();
        setError(null);// clear prev errors
        setLoading(true); // spinner

        try {
            const formDataToSend = new FormData();
            // append all fields including img
            for (let key in formData) {
                formDataToSend.append(key, formData[key]);
            }

            await signup(formDataToSend);
            navigate('/');
        } catch (err) {
            setError(err.message);
        } finally {
            setLoading(false);
        }
    };

    if (loading) return <Loading />

    return (
        <div className="FormContainer SignupForm">
            <h1>Sign Up</h1>
            <form onSubmit={handleSubmit}
                encType="multipart/form-data"
                autoComplete="off">

                {error && <p className="error" role="alert">{error}</p>}

                {/* username input */}
                <div className="form-group">
                    <label htmlFor="username">Username</label>
                    <input
                        type="text"
                        name="username"
                        id="username"
                        value={formData.username}
                        onChange={handleChange}
                        placeholder="Enter your username"
                        autoComplete="new-username"
                        required />
                </div>

                {/* password input */}
                <div className="form-group">
                    <label htmlFor="password">Password</label>
                    <input
                        type="password"
                        name="password"
                        id="password"
                        value={formData.password}
                        onChange={handleChange}
                        placeholder="Enter your password"
                        autoComplete="new-password"
                        required />
                </div>

                {/* first name input */}
                <div className="form-group">
                    <label htmlFor="firstName">First Name</label>
                    <input
                        type="text"
                        name="firstName"
                        id="firstName"
                        value={formData.firstName}
                        onChange={handleChange}
                        placeholder="Enter your first name"
                        required />
                </div>

                {/* last name input */}
                <div className="form-group">
                    <label htmlFor="lastName">Last Name</label>
                    <input
                        type="text"
                        name="lastName"
                        id="lastName"
                        value={formData.lastName}
                        onChange={handleChange}
                        placeholder="Enter your last name"
                        required />
                </div>

                {/* email input */}
                <div className="form-group">
                    <label htmlFor="email">Email</label>
                    <input
                        type="email"
                        name="email"
                        id="email"
                        value={formData.email}
                        onChange={handleChange}
                        placeholder="Enter your email"
                        required />
                </div>

                {/* birth date input */}
                <div className="form-group">
                    <label htmlFor="birthDate">Birth Date:</label>
                    <input
                        type="date"
                        name="birthDate"
                        id="birthDate"
                        value={formData.birthDate}
                        onChange={handleChange}
                        required />
                </div>

                {/* location input with autocomplete */}
                <div className="form-group">
                    <label htmlFor="location">Location</label>
                    <ReactGoogleAutocomplete
                        apiKey={import.meta.env.VITE_GOOGLE_PLACES_API_KEY}
                        onPlaceSelected={handleLocationChange}
                        options={{ types: ['(cities)'] }} // restrict to cities
                        placeholder="Type a city"
                        required />
                </div>

                {/* optional bio input */}
                <div className="form-group">
                    <label htmlFor="bio">Bio</label>
                    <textarea
                        name="bio"
                        id="bio"
                        value={formData.bio}
                        onChange={handleChange}
                        placeholder="Tell us about yourself (optional)" />
                </div>

                {/* profile picture upload */}
                <div className="form-group">
                    <label htmlFor="profilePic">Profile Picture</label>
                    <input
                        type="file"
                        name="profilePic"
                        id="profilePic"
                        onChange={handleFileUpload}
                        accept="image/*"
                    />
                </div>

                {/* pic preview */}
                {picPreview && (
                    <div className="pic-preview">
                        <img
                            src={picPreview}
                            alt="Profile Preview"
                        />
                    </div>
                )}

                {/* submit button */}
                <button type="submit" className="signup-btn">Signup</button>
            </form>
        </div>
    )
}

export default SignupForm;