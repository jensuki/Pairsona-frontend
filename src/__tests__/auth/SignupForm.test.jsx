import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { UserProvider } from '../../test_common/setupTests';
import SignupForm from '../../auth/SignupForm';
import { describe, test, expect, vi, beforeEach } from 'vitest';

// mock signup func
const mockSignup = vi.fn().mockResolvedValue({ success: true });

// helper
const renderSignupForm = () => {
    return render(
        <MemoryRouter>
            <UserProvider signup={mockSignup}>
                <SignupForm />
            </UserProvider>
        </MemoryRouter>
    );
};

// mock ReactGoogleAutocomplete input
vi.mock('react-google-autocomplete', () => ({
    default: (props) => (
        <input
            id="location"
            data-testid="autocomplete"
            required={props.required}
            placeholder={props.placeholder}
            onChange={(e) => props.onPlaceSelected({ formatted_address: e.target.value })}
        />
    ),
}));

describe('SignupForm component', () => {
    beforeEach(() => {
        mockSignup.mockClear();
    });

    test('renders without crashing', () => {
        renderSignupForm();

        const fieldLabels = [
            'Username',
            'Password',
            'First Name',
            'Last Name',
            'Email',
            'Birth Date',
            'Location',
            'Bio',
            'Profile Picture',
        ];

        // verify all labels are displayed
        fieldLabels.forEach((label) => {
            expect(screen.getByLabelText(new RegExp(label, 'i'))).toBeInTheDocument();
        });
    });

    test('matches snapshot', () => {
        const { asFragment } = renderSignupForm();
        expect(asFragment()).toMatchSnapshot();
    });

    test('handles user input correctly', () => {
        renderSignupForm();

        const usernameInput = screen.getByLabelText(/Username/i);
        fireEvent.change(usernameInput, { target: { value: 'testuser' } });

        expect(usernameInput.value).toBe('testuser');
    });

    test('handles file upload', () => {
        renderSignupForm();

        const fileInput = screen.getByLabelText(/Profile Picture/i);
        const file = new File(['test'], 'test_img.png', { type: 'image/png' });

        // simulate file upload + verify it was added
        fireEvent.change(fileInput, { target: { files: [file] } });
        expect(fileInput.files[0]).toBe(file);
        expect(fileInput.files).toHaveLength(1);
    });

    test('displays error for bio > 100 characters', () => {
        renderSignupForm();

        const bioInput = screen.getByLabelText(/Bio/i);

        // simulate exceeding valid bio length
        fireEvent.change(bioInput, {
            target: { value: 'A'.repeat(101) }
        });

        // expect error message to render
        expect(screen.getByText(/Bio cannot exceed 100 characters/i)).toBeInTheDocument();
    });

    test('handles location selection correctly', async () => {
        renderSignupForm();

        const locationInput = screen.getByPlaceholderText(/Type a city/i);
        fireEvent.change(locationInput, { target: { value: 'New York, NY' } });

        // expect location value to be properly updated
        await waitFor(() => {
            expect(locationInput.value).toBe('New York, NY');
        });
    });

    test('calls signup func on valid submission', async () => {
        renderSignupForm();

        const fieldValues = {
            Username: 'testuser',
            Password: 'password123',
            'First Name': 'Test',
            'Last Name': 'User',
            Email: 'test@test.com',
            'Birth Date': '2000-01-01',
            Bio: 'This is a test bio.',
        };

        // fill out form fields
        Object.entries(fieldValues).forEach(([label, value]) => {
            const input = screen.getByLabelText(new RegExp(label, 'i'));
            fireEvent.change(input, { target: { value } });
        });

        // simulate location selection
        const locationInput = screen.getByPlaceholderText(/Type a city/i);
        fireEvent.change(locationInput, { target: { value: 'New York, NY' } });

        // submit the form
        const submitButton = screen.getByRole('button', { name: /signup/i });
        fireEvent.click(submitButton);

        // verify signup func was called
        await waitFor(() => {
            expect(mockSignup).toHaveBeenCalled();
            // get first argument passed to mocksignup to validate content
            const formData = mockSignup.mock.calls[0][0]; // get the formdata obj
            expect(formData.get('username')).toBe('testuser');
            expect(formData.get('location')).toBe('New York, NY');
        });
    });

    test('displays error message on signup failure', async () => {
        // simulate a server error during signup
        mockSignup.mockRejectedValue(new Error('Signup Failed'));

        renderSignupForm();

        // fill out fields
        fireEvent.change(screen.getByLabelText(/username/i), { target: { value: 'testuser' } });
        fireEvent.change(screen.getByLabelText(/password/i), { target: { value: 'password123' } });
        fireEvent.change(screen.getByLabelText(/first name/i), { target: { value: 'Test' } });
        fireEvent.change(screen.getByLabelText(/last name/i), { target: { value: 'User' } });
        fireEvent.change(screen.getByLabelText(/email/i), { target: { value: 'test@test.com' } });
        fireEvent.change(screen.getByLabelText(/birth date/i), { target: { value: '2000-01-01' } });

        const locationInput = screen.getByPlaceholderText(/Type a city/i);
        fireEvent.change(locationInput, { target: { value: 'New York, NY' } });

        // submit form
        const signupBtn = screen.getByRole('button', { name: /signup/i });
        fireEvent.click(signupBtn);

        // wait for error msg to appear
        const errorMessage = await screen.findByRole('alert');
        expect(errorMessage).toHaveTextContent(/Signup Failed/i);
    });

    test('displays specific error message for invalid username', async () => {
        // mock error for invalid username field
        mockSignup.mockRejectedValue(new Error('Username must be at least 5 characters'));

        renderSignupForm();

        // fill out form with invalid username
        fireEvent.change(screen.getByLabelText(/username/i), { target: { value: 'abc' } }); // invalid
        fireEvent.change(screen.getByLabelText(/password/i), { target: { value: 'password123' } });
        fireEvent.change(screen.getByLabelText(/first name/i), { target: { value: 'Test' } });
        fireEvent.change(screen.getByLabelText(/last name/i), { target: { value: 'User' } });
        fireEvent.change(screen.getByLabelText(/email/i), { target: { value: 'test@test.com' } });
        fireEvent.change(screen.getByLabelText(/birth date/i), { target: { value: '2000-01-01' } });

        const locationInput = screen.getByPlaceholderText(/Type a city/i);
        fireEvent.change(locationInput, { target: { value: 'New York, NY' } });

        // submit
        const signupBtn = screen.getByRole('button', { name: /signup/i });
        fireEvent.click(signupBtn);

        // verify the proper error msg is rendered
        const errorMessage = await screen.findByRole('alert');
        expect(errorMessage).toHaveTextContent(/Username must be at least 5 characters/i);
    });

});
