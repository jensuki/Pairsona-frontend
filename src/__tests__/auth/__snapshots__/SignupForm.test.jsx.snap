// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SignupForm component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="FormContainer SignupForm"
  >
    <h1>
      Sign Up
    </h1>
    <form
      autocomplete="off"
      enctype="multipart/form-data"
    >
      <div
        class="form-group"
      >
        <label
          for="username"
        >
          Username
        </label>
        <input
          autocomplete="new-username"
          id="username"
          name="username"
          placeholder="Enter your username"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="password"
        >
          Password
        </label>
        <input
          autocomplete="new-password"
          id="password"
          name="password"
          placeholder="Enter your password"
          required=""
          type="password"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="firstName"
        >
          First Name
        </label>
        <input
          id="firstName"
          name="firstName"
          placeholder="Enter your first name"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="lastName"
        >
          Last Name
        </label>
        <input
          id="lastName"
          name="lastName"
          placeholder="Enter your last name"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="email"
        >
          Email
        </label>
        <input
          id="email"
          name="email"
          placeholder="Enter your email"
          required=""
          type="email"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="birthDate"
        >
          Birth Date:
        </label>
        <input
          id="birthDate"
          name="birthDate"
          required=""
          type="date"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="location"
        >
          Location
        </label>
        <input
          data-testid="autocomplete"
          id="location"
          placeholder="Type a city"
          required=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="bio"
        >
          Bio
        </label>
        <textarea
          id="bio"
          name="bio"
          placeholder="Tell us about yourself (optional)"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="profilePic"
        >
          Profile Picture
        </label>
        <input
          accept="image/*"
          id="profilePic"
          name="profilePic"
          type="file"
        />
      </div>
      <button
        class="signup-btn"
        type="submit"
      >
        Signup
      </button>
    </form>
  </div>
</DocumentFragment>
`;
